File: sample/cs/13_maze_generation_steps.cs
using System.Collections.Generic;
using System.IO;
using System;

public static class Program
{
    public static List<byte> capture(List<List<long>> grid, long w, long h, long scale)
    {
        var width = (w * scale);
        var height = (h * scale);
        var frame = Pytra.CsModule.py_runtime.py_bytearray((width * height));
        var __pytra_range_start_1 = 0;
        var __pytra_range_stop_2 = h;
        var __pytra_range_step_3 = 1;
        if (__pytra_range_step_3 == 0) throw new Exception("range() arg 3 must not be zero");
        for (var y = __pytra_range_start_1; (__pytra_range_step_3 > 0) ? (y < __pytra_range_stop_2) : (y > __pytra_range_stop_2); y += __pytra_range_step_3)
        {
            var __pytra_range_start_4 = 0;
            var __pytra_range_stop_5 = w;
            var __pytra_range_step_6 = 1;
            if (__pytra_range_step_6 == 0) throw new Exception("range() arg 3 must not be zero");
            for (var x = __pytra_range_start_4; (__pytra_range_step_6 > 0) ? (x < __pytra_range_stop_5) : (x > __pytra_range_stop_5); x += __pytra_range_step_6)
            {
                var v = (Pytra.CsModule.py_runtime.py_bool((Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(grid, y), x) == 0L)) ? 255L : 40L);
                var __pytra_range_start_7 = 0;
                var __pytra_range_stop_8 = scale;
                var __pytra_range_step_9 = 1;
                if (__pytra_range_step_9 == 0) throw new Exception("range() arg 3 must not be zero");
                for (var yy = __pytra_range_start_7; (__pytra_range_step_9 > 0) ? (yy < __pytra_range_stop_8) : (yy > __pytra_range_stop_8); yy += __pytra_range_step_9)
                {
                    var @base = ((((y * scale) + yy) * width) + (x * scale));
                    var __pytra_range_start_10 = 0;
                    var __pytra_range_stop_11 = scale;
                    var __pytra_range_step_12 = 1;
                    if (__pytra_range_step_12 == 0) throw new Exception("range() arg 3 must not be zero");
                    for (var xx = __pytra_range_start_10; (__pytra_range_step_12 > 0) ? (xx < __pytra_range_stop_11) : (xx > __pytra_range_stop_11); xx += __pytra_range_step_12)
                    {
                        Pytra.CsModule.py_runtime.py_set(frame, (@base + xx), v);
                    }
                }
            }
        }
        return Pytra.CsModule.py_runtime.py_bytes(frame);
    }

    public static void run_13_maze_generation_steps()
    {
        long cell_w = 61L;
        long cell_h = 45L;
        long scale = 4L;
        string out_path = "sample/out/13_maze_generation_steps.gif";
        var start = Pytra.CsModule.time.perf_counter();
        List<List<long>> grid = new List<List<long>> {  };
        var __pytra_range_start_13 = 0;
        var __pytra_range_stop_14 = cell_h;
        var __pytra_range_step_15 = 1;
        if (__pytra_range_step_15 == 0) throw new Exception("range() arg 3 must not be zero");
        for (var __pytra_unused_16 = __pytra_range_start_13; (__pytra_range_step_15 > 0) ? (__pytra_unused_16 < __pytra_range_stop_14) : (__pytra_unused_16 > __pytra_range_stop_14); __pytra_unused_16 += __pytra_range_step_15)
        {
            List<long> row = new List<long> {  };
            var __pytra_range_start_17 = 0;
            var __pytra_range_stop_18 = cell_w;
            var __pytra_range_step_19 = 1;
            if (__pytra_range_step_19 == 0) throw new Exception("range() arg 3 must not be zero");
            for (var __pytra_unused_20 = __pytra_range_start_17; (__pytra_range_step_19 > 0) ? (__pytra_unused_20 < __pytra_range_stop_18) : (__pytra_unused_20 > __pytra_range_stop_18); __pytra_unused_20 += __pytra_range_step_19)
            {
                Pytra.CsModule.py_runtime.py_append(row, 1L);
            }
            Pytra.CsModule.py_runtime.py_append(grid, row);
        }
        List<Tuple<long, long>> stack = new List<Tuple<long, long>> { Tuple.Create(1L, 1L) };
        Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, 1L), 1L, 0L);
        List<Tuple<long, long>> dirs = new List<Tuple<long, long>> { Tuple.Create(2L, 0L), Tuple.Create((-2L), 0L), Tuple.Create(0L, 2L), Tuple.Create(0L, (-2L)) };
        List<List<byte>> frames = new List<List<byte>> {  };
        long step = 0L;
        while (Pytra.CsModule.py_runtime.py_bool((Pytra.CsModule.py_runtime.py_len(stack) > 0L)))
        {
            var last_index = (Pytra.CsModule.py_runtime.py_len(stack) - 1L);
            var __pytra_tuple_21 = Pytra.CsModule.py_runtime.py_get(stack, last_index);
            var x = __pytra_tuple_21.Item1;
            var y = __pytra_tuple_21.Item2;
            List<Tuple<long, long, long, long>> candidates = new List<Tuple<long, long, long, long>> {  };
            var __pytra_range_start_22 = 0;
            var __pytra_range_stop_23 = 4L;
            var __pytra_range_step_24 = 1;
            if (__pytra_range_step_24 == 0) throw new Exception("range() arg 3 must not be zero");
            for (var k = __pytra_range_start_22; (__pytra_range_step_24 > 0) ? (k < __pytra_range_stop_23) : (k > __pytra_range_stop_23); k += __pytra_range_step_24)
            {
                var __pytra_tuple_25 = Pytra.CsModule.py_runtime.py_get(dirs, k);
                var dx = __pytra_tuple_25.Item1;
                var dy = __pytra_tuple_25.Item2;
                var nx = (x + dx);
                var ny = (y + dy);
                if (Pytra.CsModule.py_runtime.py_bool(((nx >= 1L) && (nx < (cell_w - 1L)) && (ny >= 1L) && (ny < (cell_h - 1L)) && (Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(grid, ny), nx) == 1L))))
                {
                    if (Pytra.CsModule.py_runtime.py_bool((dx == 2L)))
                    {
                        Pytra.CsModule.py_runtime.py_append(candidates, Tuple.Create(nx, ny, (x + 1L), y));
                    }
                    else
                    {
                        if (Pytra.CsModule.py_runtime.py_bool((dx == (-2L))))
                        {
                            Pytra.CsModule.py_runtime.py_append(candidates, Tuple.Create(nx, ny, (x - 1L), y));
                        }
                        else
                        {
                            if (Pytra.CsModule.py_runtime.py_bool((dy == 2L)))
                            {
                                Pytra.CsModule.py_runtime.py_append(candidates, Tuple.Create(nx, ny, x, (y + 1L)));
                            }
                            else
                            {
                                Pytra.CsModule.py_runtime.py_append(candidates, Tuple.Create(nx, ny, x, (y - 1L)));
                            }
                        }
                    }
                }
            }
            if (Pytra.CsModule.py_runtime.py_bool((Pytra.CsModule.py_runtime.py_len(candidates) == 0L)))
            {
                Pytra.CsModule.py_runtime.py_pop(stack);
            }
            else
            {
                var sel = Pytra.CsModule.py_runtime.py_get(candidates, ((((x * 17L) + (y * 29L)) + (Pytra.CsModule.py_runtime.py_len(stack) * 13L)) % Pytra.CsModule.py_runtime.py_len(candidates)));
                var __pytra_tuple_26 = sel;
                var nx = __pytra_tuple_26.Item1;
                var ny = __pytra_tuple_26.Item2;
                var wx = __pytra_tuple_26.Item3;
                var wy = __pytra_tuple_26.Item4;
                Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, wy), wx, 0L);
                Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, ny), nx, 0L);
                Pytra.CsModule.py_runtime.py_append(stack, Tuple.Create(nx, ny));
            }
            if (Pytra.CsModule.py_runtime.py_bool(((step % 25L) == 0L)))
            {
                Pytra.CsModule.py_runtime.py_append(frames, capture(grid, cell_w, cell_h, scale));
            }
            step = (step + 1L);
        }
        Pytra.CsModule.py_runtime.py_append(frames, capture(grid, cell_w, cell_h, scale));
        Pytra.CsModule.gif_helper.save_gif(out_path, (cell_w * scale), (cell_h * scale), frames, Pytra.CsModule.gif_helper.grayscale_palette(), delay_cs: 4L, loop: 0L);
        var elapsed = (Pytra.CsModule.time.perf_counter() - start);
        Pytra.CsModule.py_runtime.print("output:", out_path);
        Pytra.CsModule.py_runtime.print("frames:", Pytra.CsModule.py_runtime.py_len(frames));
        Pytra.CsModule.py_runtime.print("elapsed_sec:", elapsed);
    }

    public static void Main(string[] args)
    {
        run_13_maze_generation_steps();
    }
}

File: sample/cpp/13_maze_generation_steps.cpp
#include "cpp_module/py_runtime.h"




list<uint8> capture(list<list<int64>> grid, int64 w, int64 h, int64 scale) {
    int64 v;
    int64 base;
    
    int64 width = w * scale;
    int64 height = h * scale;
    list<uint8> frame = list<uint8>(width * height);
    for (int64 y = 0; y < h; ++y) {
        for (int64 x = 0; x < w; ++x) {
            v = (grid[y][x] == 0 ? 255 : 40);
            for (int64 yy = 0; yy < scale; ++yy) {
                base = (y * scale + yy) * width + x * scale;
                for (int64 xx = 0; xx < scale; ++xx)
                    frame[base + xx] = v;
            }
        }
    }
    return list<uint8>(frame);
}

void run_13_maze_generation_steps() {
    int64 x;
    int64 y;
    list<std::tuple<int64, int64, int64, int64>> candidates;
    int64 dx;
    int64 dy;
    int64 nx;
    int64 ny;
    std::tuple<int64, int64, int64, int64> sel;
    int64 wx;
    int64 wy;
    
    int64 cell_w = 89;
    int64 cell_h = 67;
    int64 scale = 5;
    int64 capture_every = 20;
    str out_path = "sample/out/13_maze_generation_steps.gif";
    
    float64 start = perf_counter();
    list<list<int64>> grid = [&]() -> list<list<int64>> {     list<list<int64>> __out;     for (int64 _ = 0; (_ < cell_h); _ += (1)) {         __out.push_back(py_repeat(list<int64>{1}, cell_w));     }     return __out; }();
    list<std::tuple<int64, int64>> stack = list<std::tuple<int64, int64>>{std::make_tuple(1, 1)};
    grid[1][1] = 0;
    
    list<std::tuple<int64, int64>> dirs = list<std::tuple<int64, int64>>{std::make_tuple(2, 0), std::make_tuple(-2, 0), std::make_tuple(0, 2), std::make_tuple(0, -2)};
    list<list<uint8>> frames = list<list<uint8>>{};
    int64 step = 0;
    
    while ((py_len(stack) != 0)) {
        auto __tuple_1 = py_at(stack, -1);
        x = std::get<0>(__tuple_1);
        y = std::get<1>(__tuple_1);
        candidates = list<std::tuple<int64, int64, int64, int64>>{};
        for (int64 k = 0; k < 4; ++k) {
            auto __tuple_2 = dirs[k];
            dx = std::get<0>(__tuple_2);
            dy = std::get<1>(__tuple_2);
            nx = x + dx;
            ny = y + dy;
            if (nx >= 1 && nx < cell_w - 1 && ny >= 1 && ny < cell_h - 1 && grid[ny][nx] == 1) {
                if (dx == 2) {
                    candidates.push_back(std::make_tuple(nx, ny, x + 1, y));
                } else {
                    if (dx == -2) {
                        candidates.push_back(std::make_tuple(nx, ny, x - 1, y));
                    } else {
                        if (dy == 2)
                            candidates.push_back(std::make_tuple(nx, ny, x, y + 1));
                        else
                            candidates.push_back(std::make_tuple(nx, ny, x, y - 1));
                    }
                }
            }
        }
        
        if (py_len(candidates) == 0) {
            py_pop(stack);
        } else {
            sel = candidates[(x * 17 + y * 29 + py_len(stack) * 13) % py_len(candidates)];
            auto __tuple_3 = sel;
            nx = std::get<0>(__tuple_3);
            ny = std::get<1>(__tuple_3);
            wx = std::get<2>(__tuple_3);
            wy = std::get<3>(__tuple_3);
            grid[wy][wx] = 0;
            grid[ny][nx] = 0;
            stack.push_back(std::make_tuple(nx, ny));
        }
        
        if (step % capture_every == 0)
            frames.push_back(capture(grid, cell_w, cell_h, scale));
        step++;
    }
    
    frames.push_back(capture(grid, cell_w, cell_h, scale));
    save_gif(out_path, cell_w * scale, cell_h * scale, frames, grayscale_palette(), 4, 0);
    float64 elapsed = perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", py_len(frames));
    py_print("elapsed_sec:", elapsed);
}

int main() {
    run_13_maze_generation_steps();
    return 0;
}
